examples/classification_example/main_train.py

```diff
@@ -0,0 +1,238 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+import numpy as np
+import torch
+import yaml
+
+from torch import nn
+from torch.optim import SGD
+from torch.optim.optimizer import Optimizer
+from texar.torch.data import Batch
+from tqdm import tqdm
+from typing import Iterator, Dict
+from forte.common.configuration import Config
+from forte.data.extractor.attribute_extractor \
+    import AttributeExtractor
+from forte.data.extractor.base_extractor \
+    import BaseExtractor
+from forte.train_preprocessor import TrainPreprocessor
+from forte.data.readers.imdb_reader import IMDBReader
+from forte.pipeline import Pipeline
+from ft.onto.base_ontology import Sentence, Token
+from texar.torch.modules.embedders import WordEmbedder
+from examples.classification_example.cnn import CNN_Classifier
+from forte.processors.base.data_augment_processor import ReplacementDataAugmentProcessor
+from texar.torch.modules.classifiers.bert_classifier import BERTClassifier
+from examples.classification_example.util import pad_each_bach
+
+logger = logging.getLogger(__name__)
+
+logging.basicConfig(level=logging.INFO)
+
+device = torch.device("cuda") if torch.cuda.is_available() \
+    else torch.device("cpu")
+
+
+def construct_word_embedding_table(embed_dict, extractor: BaseExtractor):
+    embedding_dim = list(embed_dict.values())[0].shape[-1]
+
+    scale = np.sqrt(3.0 / embedding_dim)
+    table = np.empty(
+        [extractor.size(), embedding_dim], dtype=np.float32
+    )
+    oov = 0
+    for word, index in extractor.items():
+        if word in embed_dict:
+            embedding = embed_dict[word]
+        elif word.lower() in embed_dict:
+            embedding = embed_dict[word.lower()]
+        else:
+            embedding = np.random.uniform(
+                -scale, scale, [1, embedding_dim]
+            ).astype(np.float32)
+            oov += 1
+        table[index, :] = embedding
+    return torch.from_numpy(table)
+
+
+def create_model(schemes: Dict[str, Dict[str, BaseExtractor]],
+                 config: Config, in_channels: int):
+    text_extractor: BaseExtractor = schemes["text_tag"]["extractor"]
+    embedding_dict = {}
+    fake_tensor = torch.tensor([0.0 for i in range(100)])
+    for word, index in text_extractor.items():
+        embedding_dict[word] = fake_tensor
+
+    word_embedding_table = \
+        construct_word_embedding_table(embedding_dict, text_extractor)
+
+    model: nn.Module = \
+        CNN_Classifier(in_channels=in_channels,
+                       word_embedding_table=word_embedding_table)
+
+    if config.config_model.model == "bert":
+        model: nn.Module = BERTClassifier()
+
+    return model, word_embedding_table
+
+
+def train(model: nn.Module, optim: Optimizer, batch: Batch, max_sen_length: int):
+    word = batch["text_tag"]["data"]
+    labels = batch["label_tag"]["data"]
+    optim.zero_grad()
+
+    logits, pred = None, None
+
```

Comment:
I am very curious about calling model forward part here as both `cnn` and `BERT` does not need to use any of mask to get the actual length of each sentence?

Reply:
I think you need to pass in the max length as mask for CNN: https://github.com/asyml/texar-pytorch/blob/787311e1b51462fada6dc7359868b6c976dad502/texar/torch/modules/classifiers/conv_classifiers.py#L166

Reply:
They only need the max_sentence_length. The in_channels in the cnn input is the max_sentence_length.

Reply:
But from the documentation, you have to feed `max_sentence_length` to the `forward` step, is it?

Reply:
In the forward function, the input matrix is [batch_size][max_sentence_length][word_vec_dim]
I store the max_sentence_length here.
https://github.com/ziqian98/forte/blob/lzq_new_classification/examples/classification_example/cnn.py#L45

Reply:
So you pad yourself before pushing in to forward.

But actually, it seems cleaner to me if you do:

`self.classifier(word_pad_embed, max_sen_len)`

You don't even need to do `word_pad = pad_each_bach(word, self.max_sen_len)` right?

Reply:
Since there is no max_sentence_length in conv_classifier forward step(https://github.com/asyml/texar-pytorch/blob/master/texar/torch/modules/classifiers/conv_classifiers.py#L181). It only needs word_pad_embed to forward.

And actually, word_pad_embed is  [batch_size][max_sentence_length][word_vec_dim]. This matrix contains the information of max_sentence_length. So we do not need to pass max_sentence_length in self.classifier.

I think the original author in Texar conv_classifier assume that the input matrix is already padded outside the forward so that he did not use a max_sen_length as an input to pad in the forward.

Reply:
Sorry I was originally talking about the masks and somehow confused myself and talk about max length. I agree you have to pad it outside though.

However, I think we have to miss the masks? https://github.com/asyml/texar-pytorch/blob/master/texar/torch/modules/classifiers/conv_classifiers.py#L183

Reply:
Yes, but this field is optional. CNN model does not need mask originally. Do we still need to add this mask?

Reply:
Probably won't hurt. But that really depend on your evaluation results.

Reply:
Okay, I will record this and do some experiments to investigate this.

Reply:
Optional does not necessarily mean we do not need them. I am not quite sure about CNN. But for BERT, it is essentially a Transformer and Transformer definitely need some masks to know which token is the actual word and which token is `<PAD>`. Why BERT does not need this? I guess the parameter `sequence_length` just served this purpose?

Reply:
Yes you are right. sequence_length in the bert_classifier forward function is the mask. The author set this field to optional.

Reply:
A mask is already added.